From a63d42ea42bf42e6a8fcb09ce9e462bc2f9a80ca Mon Sep 17 00:00:00 2001
From: Antonin Bas <abas@vmware.com>
Date: Wed, 7 Jul 2021 09:55:43 -0700
Subject: [PATCH] Fix issues in p4dbg.py after switch to Python3 (#1017)

Fixes #1016
---
 tools/p4dbg.py | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/tools/p4dbg.py b/tools/p4dbg.py
index dba99e6..a88fb03 100755
--- a/tools/p4dbg.py
+++ b/tools/p4dbg.py
@@ -260,7 +260,7 @@ class Msg(object):
         return 0
 
     def generate(self):
-        return []
+        return b""
 
     def __str__(self):
         return ""
@@ -287,8 +287,10 @@ def make_extract_function(P, fmt):
                     v_str = 'Empty'
                     v_int = 0
                 else:
-                    v_str = ':'.join(x.encode('hex') for x in v)
-                    v_int = int(v.encode('hex'), 16)
+                    # only available in Python 3.8 :(
+                    # v_str = v.hex(':')
+                    v_str = ':'.join(['{:02x}'.format(x) for x in v])
+                    v_int = int(v.hex(), 16)
                 setattr(self, name + "_str", v_str)
                 setattr(self, name + "_int", v_int)
         return offset
@@ -322,9 +324,7 @@ def make_generate_function(P, fmt):
                 assert(src is None)
                 assert(type(uf) is str)
                 f = uf
-            # encoding problem if using string
-            # nnpy expects unicode, which does not roll
-            s += list(struct.pack(f, v))
+            s += struct.pack(f, v)
         return s
 
     return generate
@@ -390,10 +390,10 @@ def _test():
     m = Msg_FieldValue()
     # < required to prevent 8-byte alignment
     req = struct.pack("<iQQQQQi", MsgType.FIELD_VALUE, 0, 99, 1, 2, 55, 3)
-    req += struct.pack(">3s", "\x01\x02\x03")
+    req += struct.pack(">3s", b"\x01\x02\x03")
     m.extract(req)
     s = m.generate()
-    assert(s == list(req))
+    assert(s == req)
 
 _test()
 
@@ -1206,6 +1206,7 @@ def main():
         print(e)
         if c.attached:
             c.say_bye()
+        raise
 
 if __name__ == "__main__":
     main()
-- 
2.33.1

